libs/langchain/langchain/schema/runnable/history.py

```diff
@@ -0,0 +1,254 @@
+            MessagesOrDictWithMessages,
+            Union[str, BaseMessage, MessagesOrDictWithMessages],
+        ],
+        factory: Callable[[str], BaseChatMessageHistory],
+        *,
+        input_messages_key: Optional[str] = None,
+        output_messages_key: Optional[str] = None,
+        message_history_key: Optional[str] = None,
+        **kwargs: Any,
+    ) -> None:
+        """Initialize RunnableWithMessageHistory.
+
+        Args:
+            runnable: The base Runnable to be wrapped.
+
+                Must take as input one of:
+                - A sequence of BaseMessages
+                - A dict with one key for all messages
+                - A dict with one key for the current input string/message(s) and
+                    a separate key for historical messages. If the input key points
+                    to a string, it will be treated as a HumanMessage in history.
+
+                Must return as output one of:
+                - A string which can be treated as an AIMessage
+                - A BaseMessage or sequence of BaseMessages
+                - A dict with a key for a BaseMessage or sequence of BaseMessages
+
+            factory: Function that returns a new BaseChatMessageHistory given a
+                session id. Should take a single parameter `session_id` which is a
+                string.
+            input_messages_key: Must be specified if the base runnable accepts a dict
+                as input.
+            output_messages_key: Must be specified if the base runnable returns a dict
+                as output.
+            message_history_key: Must be specified if the base runnable accepts a dict
+                as input and expects a separate key for historical messages.
+            **kwargs: Arbitrary additional kwargs to pass to parent class
+                ``RunnableBindingBase`` init.
+        """
+        history_chain: Runnable = RunnableLambda(
+            self._enter_history, self._aenter_history
+        ).with_config(run_name="load_history")
+        messages_key = message_history_key or input_messages_key
+        if messages_key:
+            history_chain = RunnablePassthrough.assign(
+                **{messages_key: history_chain}
+            ).with_config(run_name="insert_history")
+        bound = (
+            history_chain | runnable.with_listeners(on_end=self._exit_history)
+        ).with_config(run_name="RunnableWithMessageHistory")
+        super().__init__(
+            factory=factory,
+            input_messages_key=input_messages_key,
+            output_messages_key=output_messages_key,
+            bound=bound,
+            message_history_key=message_history_key,
+            **kwargs,
+        )
+
+    @property
+    def config_specs(self) -> List[ConfigurableFieldSpec]:
+        return get_unique_config_specs(
+            super().config_specs
+            + [
+                ConfigurableFieldSpec(
+                    id="user_id",
```

Comment:
What's the reason why we need both instead of just having a `session_id` and having the client decide how to deal with that? The reason is that we're opinionated about it's assembled (i.e., we're not passing both parameters independently)

Reply:
cc @nfcampos

Reply:
This is what we wanted for opengpts doesn't necessarily make sense here

Reply:
With that said I'm OK with user id and thread id as being the default parametrization, but another thing we could do is to accept the parameterization through the init, and pass it along to the factory